File: webApp/src/practice/grading/multipleChoice.test.ts
import { describe, it, expect } from 'vitest';
import { buildChoices } from './multipleChoice';
import type { FlashcardDto } from '../../api/types';

const deck: FlashcardDto[] = [
  { question: 'France', answer: 'Paris' },
  { question: 'Japan', answer: 'Tokyo' },
  { question: 'Italy', answer: 'Rome' },
  { question: 'Spain', answer: 'Madrid' },
  { question: 'Germany', answer: 'Berlin' },
];

// Deterministic RNG (always 0) → a fixed shuffle, so results are stable across runs.
const fixed = () => 0;

describe('buildChoices', () => {
  it('returns up to 4 options including the correct answer', () => {
    const choices = buildChoices(deck[0], deck, 4, fixed);
    expect(choices.length).toBe(4);
    expect(choices).toContain('Paris');
  });

  it('draws unique distractors from other cards, never the correct answer', () => {
    const choices = buildChoices(deck[0], deck, 4, fixed);
    const distractors = choices.filter((c) => c !== 'Paris');
    expect(distractors).toHaveLength(3);
    for (const d of distractors) {
      expect(['Tokyo', 'Rome', 'Madrid', 'Berlin']).toContain(d);
    }
    expect(new Set(choices).size).toBe(choices.length); // no duplicates
  });

  it('de-duplicates case-insensitively and skips blank answers', () => {
    const dupes: FlashcardDto[] = [
      { question: 'A', answer: 'Paris' }, // correct
      { question: 'B', answer: 'paris' }, // duplicate of correct (case)
      { question: 'C', answer: '   ' }, // blank
      { question: 'D', answer: 'Tokyo' },
      { question: 'E', answer: 'tokyo' }, // duplicate distractor
    ];
    const choices = buildChoices(dupes[0], dupes, 4, fixed);
    expect(choices).toContain('Paris');
    expect(choices.filter((c) => c !== 'Paris')).toEqual(['Tokyo']);
  });

  it('yields fewer than 4 options for a small deck', () => {
    const small: FlashcardDto[] = [
      { question: 'France', answer: 'Paris' },
      { question: 'Japan', answer: 'Tokyo' },
    ];
    expect(buildChoices(small[0], small, 4, fixed).slice().sort()).toEqual(['Paris', 'Tokyo']);
  });

  it('is deterministic with a seeded RNG', () => {
    expect(buildChoices(deck[0], deck, 4, fixed)).toEqual(buildChoices(deck[0], deck, 4, fixed));
  });
});

File: shared/src/commonTest/kotlin/com/rrbrambley/flashcards/practice/grading/MultipleChoiceTest.kt
package com.rrbrambley.flashcards.practice.grading

import com.rrbrambley.flashcards.shared.domain.Flashcard
import kotlin.random.Random
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals

class MultipleChoiceTest {

    private val deck = listOf(
        Flashcard("France", "Paris"),
        Flashcard("Japan", "Tokyo"),
        Flashcard("Italy", "Rome"),
        Flashcard("Spain", "Madrid"),
        Flashcard("Germany", "Berlin"),
    )

    // A fresh seeded RNG per call so the shuffle is deterministic.
    private fun seeded() = Random(42)

    @Test
    fun returnsUpToFourOptionsIncludingTheCorrectAnswer() {
        val choices = buildChoices(deck[0], deck, count = 4, random = seeded())
        assertEquals(4, choices.size)
        assertContains(choices, "Paris")
    }

    @Test
    fun drawsUniqueDistractorsFromOtherCardsNeverTheCorrectAnswer() {
        val choices = buildChoices(deck[0], deck, count = 4, random = seeded())
        val distractors = choices.filter { it != "Paris" }
        assertEquals(3, distractors.size)
        distractors.forEach { assertContains(listOf("Tokyo", "Rome", "Madrid", "Berlin"), it) }
        assertEquals(choices.size, choices.toSet().size) // no duplicates
    }

    @Test
    fun deduplicatesCaseInsensitivelyAndSkipsBlankAnswers() {
        val dupes = listOf(
            Flashcard("France", "Paris"),
            Flashcard("Capital of France", "paris"), // case-dup of the correct answer → dropped
            Flashcard("Japan", "Tokyo"),
            Flashcard("Mystery", ""), // blank → dropped
        )
        val choices = buildChoices(dupes[0], dupes, count = 4, random = seeded())
        assertContains(choices, "Paris")
        assertEquals(listOf("Tokyo"), choices.filter { it != "Paris" })
    }

    @Test
    fun yieldsFewerThanCountForASmallDeck() {
        val small = listOf(Flashcard("France", "Paris"), Flashcard("Japan", "Tokyo"))
        assertEquals(listOf("Paris", "Tokyo"), buildChoices(small[0], small, count = 4, random = seeded()).sorted())
    }

    @Test
    fun isDeterministicWithASeededRng() {
        assertEquals(
            buildChoices(deck[0], deck, count = 4, random = Random(7)),
            buildChoices(deck[0], deck, count = 4, random = Random(7)),
        )
    }
}
